test/inductor/test_fp8.py

```diff
@@ -346,6 +384,298 @@ def ln_fp8(x: Tensor, scale: Tensor, amax_buffer: Tensor):
         )
 
 
+@instantiate_parametrized_tests
+class TestFP8Lowering(TestCase):
+    @unittest.skipIf(TEST_WITH_ROCM, "FP8 is not supported on ROCM")
+    @unittest.skipIf(not SM90OrLater, "FP8 is only supported on H100+")
+    @parametrize("dtype", (torch.bfloat16, torch.float32))
+    @parametrize("shape", ("16,16,32", "1024,1024,512"))
+    @parametrize("has_bias", (False, True))
+    @parametrize("use_fast_accum", (False, True))
+    def test_tensorwise_scaling(
+        self, dtype: torch.dtype, shape: str, has_bias: bool, use_fast_accum: bool
+    ):
+        if dtype is torch.float32 and has_bias:
+            self.skipTest("bias is not supported when output dtype is float32")
+
+        device = "cuda"
+        dtype_float8 = torch.float8_e4m3fn
+
+        shape = [int(dim) for dim in shape.split(",")]
+        M, K, N = shape  # Matmul Y = X [M, K] x W [N, K]
+        # input and output dtypes of _scaled_mm do not need to be the same, but
+        # typically in a model they are
+        x = torch.rand(M, K, dtype=dtype, device=device)
+        w = torch.rand(N, K, dtype=dtype, device=device)
+        bias = None
+        if has_bias:
+            bias = torch.rand(N, device=device, dtype=torch.bfloat16)
+
+        # quantize weight (prior to inference)
+        w_fp8, w_inverse_scale = _quantize_tensorwise(w, dtype_float8)
+        w_t_fp8 = w_fp8.t()
+
+        def linear(x, w_t_fp8, w_inverse_scale, bias):
+            # quantize input x
+            x_fp8, x_inverse_scale = _quantize_tensorwise(x, dtype_float8)
+            y = torch._scaled_mm(
+                x_fp8,
+                w_t_fp8,
+                x_inverse_scale,
+                w_inverse_scale,
+                bias,
+                out_dtype=dtype,
+                use_fast_accum=use_fast_accum,
+            )
+            return y
+
+        y_eager = linear(
+            x,
+            w_t_fp8,
+            w_inverse_scale,
+            bias,
+        )
+        linear_compiled = torch.compile(linear, backend="inductor", mode="max-autotune")
+        y_compiled = linear_compiled(
+            x,
+            w_t_fp8,
+            w_inverse_scale,
+            bias,
+        )
+        self.assertEqual(y_eager.dtype, dtype)
+        self.assertEqual(y_compiled.dtype, dtype)
+        torch.testing.assert_close(y_eager, y_compiled, rtol=5e-2, atol=0)
```

Comment:
the tolerance is higher than I would have expected.  If we compare, for example, the cuBLAS kernel (via torch._scaled_mm) versus the triton fp8 tutorial's matmul, is the tolerance similar?

Reply:
I see one issue is that I include input quantization inside the compiled function, making the difference greater than if only the matmul is compiled. Updated the UTs to only include matmul in the compiled function, and all `rtol` to 0.01.

I also realized that I made the mistake of using `rand()` (Uniform[0, 1)) instead of `randn()` (Normal(0, 1)) to generate input matrices. Using `randn()`, some result elements are close to 0, so the relative difference can be quite large (for M, K, N = 2, 1024, 2048, one element is -0.0432 in the compiled case and -0.0234 via the cuBLAS kernel, 80% relative difference). So I've increased the `atol` to 0.05 or 0.07 to allow for this (see below on justification). 

> If we compare, for example, the cuBLAS kernel (via torch._scaled_mm) versus the triton fp8 tutorial's matmul, is the tolerance similar?

The Triton template used for the compiled case follows that of the mm kernel (with scaling and bias added), which is expressed slightly differently than the [tutorial's matmul](https://triton-lang.org/main/getting-started/tutorials/03-matrix-multiplication.html#sphx-glr-getting-started-tutorials-03-matrix-multiplication-py) but the approach is exactly the same. When I compared the cuBLAS kernel, tutorial's Triton and the templated Triton kernels, I did see larger relative difference (for element close to 0) between the templated Triton and cuBLAS for some shapes e.g. (3, 1024, 2048), (2, 1024, 2048. This turned out to be because certain configs (block M size etc) led to different results (expected as float addition is not associative - see my post in the internal Triton group), and once we use the same config for the tutorial's Triton kernel, then both Triton kernels had the same relative difference to cuBLAS. The difference is very small compared to how imprecise they are from the bfloat16 result. Repro script and output: P1496984171.